packages/promise-tracker/lib/index.ts

```diff
@@ -0,0 +1,104 @@
+const _outstandingPromiseInstanceMap = new WeakMap<Object, Set<Promise<any>>>();
+const _outstandingPromiseMap = new Map<Promise<any>, boolean>();
+
+export function tracked<T extends Object>(
+  target: T,
+  key: string | Symbol,
+  descriptor: PropertyDescriptor
+) {
+  if (typeof descriptor.value === "function") {
+    const impl = descriptor.value;
+
+    // wrapper function that keeps a list of outstanding promises for each
+    // instance of DashboardMessageBusClient
+    function newImpl(...args: any[]) {
+      const returnValue = impl.apply(this, args);
+
+      // only track async functions
+      if (returnValue.then && typeof returnValue.then === "function") {
+        const trackedTask = returnValue
+          .then((result: any) => {
+            _cleanUpTrackedTask(this, trackedTask);
+            return result;
+          })
+          .catch((err: any) => {
+            _cleanUpTrackedTask(this, trackedTask);
+            throw err;
+          });
+
+        _trackTask(this, trackedTask);
+        return trackedTask;
+      }
+
+      return returnValue;
+    }
+
+    return {
+      ...descriptor,
+      value: newImpl
+    };
+  }
+
+  return descriptor;
+}
+
+export interface WaitForOutstandingPromiseOptions<T> {
+  catchHandler?: (err?: any) => void;
+  target?: T;
+}
+
+export async function waitForOutstandingPromises<T extends Object>(
+  options?: WaitForOutstandingPromiseOptions<T>
+) {
+  const { target } = options || {};
+  let { catchHandler } = options || {};
+
+  const iterable =
+    (target
+      ? _outstandingPromiseInstanceMap.get(target)?.values()
+      : _outstandingPromiseMap.keys()) || [];
+
+  catchHandler = catchHandler || (() => {});
+
+  await Promise.all(Array.from(iterable, p => p.catch(catchHandler)));
```

Comment:
I think this should be `Promise.allSettled`, as `Promise.all` may resolve as soon as a promise throws, and since `catchHandler` can be supplied by the caller it can throw and cause an early exit.

Reply:
Agreed re: `Promise.allSettled` I might need to shim that for compatibility, however. But I'll look into it.

> What happens if new tracked promises are started after `await waitForOutstandingPromises()` is called?

`waitForOutstandingPromises` is meant to be called immediately prior to terminating the process via `process.exit`. Honestly it's kind of a crap workaround for the fact that we sometimes have some handles left open at the end of a CLI run. Primarily we want to ensure that anything firing off to the message bus has had enough time to succeed, but we don't want to depend on a static delay (e.g. `await delay(1000); process.exit(exitCode);`).

A more ideal choice would be to ensure that there aren't any open handles left dangling when the CLI finishes executing, but this was proving rather difficult to debug.

Reply:
Actually now that I look at this again I remembered that effectively removed the need for a shim for `Promise.allSettled`. I even checked for this problem w/ [a test](https://github.com/trufflesuite/truffle/pull/4936/files/9f79abcd47364dcb729bccdd51ec5557dd798e25#diff-cf37491629209eac850cb0b88a1914fee649ef2152aebdcd22cd35a17de3e2fbR83-R97).

Reply:
I tried `allSettled` but the existing implementation is effectively equivalent. With the existing implementation I map the catch handler to the individual promises during aggregation.

To preserve the property where `catchHandler` is called once per rejected promise, I'd need to map the `catchHandler` to the returned array of status objects, filtering on those with `result.status === "rejected"`.

The current implementation achieves the same behaviour for the caller in a single line.